[PyTorch][ET] Collect Process Groups Mapping Info

louisfeng opened the pull request:
Summary: Add the logics and interface to log ProcessGroup comms configuration (unique ID, type, and ranks info).

Test Plan:
Testing in HPC:
```
TORCH_LOGS=all ../buck-out/v2/gen/fbcode/c8344b52091f4f7f/hpc/models/ads/__ads_10x_launcher__/ads_10x_launcher.par  +launcher=local launcher.num_trainers=4 +data_loader=random data_loader.num_batches=2000
```
Example output in ET:
```
    {
      "name": "## process_group:init ##", "id": 3, "rf_id": 1, "parent": 2, "fw_parent": 0, "seq_id": -1, "scope": 7, "tid": 1, "fw_tid": 0, "op_schema": "",
      "inputs": ["[{'pg_id': 140538064364672, 'backend_id': 140538060772480, 'backend_config': 'cuda:nccl', 'ranks': {0: 0, 1: 1, 2: 2, 3: 3}}, {'pg_id': 140538064363904, 'backend_id': 140538042628864, 'backend_config': 'cuda:nccl', 'ranks': {0: 0, 1: 1, 2: 2, 3: 3}}]"], "input_shapes": [[]], "input_types": ["String"],
      "outputs": [], "output_shapes": [], "output_types": []
    },
```

Differential Revision: D46321690

kwen2501 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1375,6 +1375,11 @@ that adds a prefix to each key inserted to the store.
           .def("rank", &::c10d::ProcessGroup::getRank)
           .def("size", &::c10d::ProcessGroup::getSize)
           .def("name", &::c10d::ProcessGroup::getBackendName)
+          .def("id", &::c10d::ProcessGroup::getID)
+          .def(
+              "backend_id",
+              &::c10d::ProcessGroup::getBackendID,
+              py::arg("backend_type"))
```
I wonder if these APIs must be exposed to end users?

louisfeng replied:
Added prefix.

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -513,6 +520,22 @@ def pg_coalesce_state(self) -> Dict[ProcessGroup, List[Union[_CollOp, P2POp]]]:
         return self._pg_default_device
 
+    @property
+    def pg_config_info(self) -> List[Dict[str, str]]:
```
Where is this function used?

louisfeng replied:
This is currently used in one of the trace collection step to inject this information into the trace so the traces can be used for network replay and triaging.

kwen2501 replied:
So this intends to be a user-facing API. Would you mind adding a short description under the def line? Thank you!

louisfeng replied:
I have added comments to clarify the usage of this function.

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -513,6 +520,22 @@ def pg_coalesce_state(self) -> Dict[ProcessGroup, List[Union[_CollOp, P2POp]]]:
+    def pg_config_info(self) -> List[Dict[str, str]]:
+        config_info = []
+        for pg, backend in self.pg_map.items():
+            backend_type = Backend.backend_type_map[backend[0]]
```
why backend[0]? Some clarification would be helpful if you think [0] is the right behavior.

louisfeng replied:
I'll add some comments. `backend` is a tuple with the first element being the backend type (nccl, etc.).

louisfeng replied:
Added comments.

kwen2501 commented on torch/_C/_autograd.pyi:
```diff
@@ -91,7 +91,7 @@ def _profiler_enabled() -> bool: ...
 def _kineto_step() -> None: ...
 def kineto_available() -> bool: ...
-def _record_function_with_args_enter(name: str, args: List[Any]) -> torch.Tensor: ...
+def _record_function_with_args_enter(name: str, *args) -> torch.Tensor: ...
```
How is this change relate to the rest of the PR? (Is it necessary?)

louisfeng replied:
This is required for the internal diff (which has other code depends on this external diff). This fixes an edge case where `args` is a single value, which would work correctly with `*args` but not `args: List[Any]`. A workaround is to put the single value in a list, but that would change its type. `List[Any]` is a mismatch of `py::args args` semantics (see https://pybind11.readthedocs.io/en/stable/advanced/functions.html#accepting-args-and-kwargs)

kwen2501 reviewed (approved):
Thanks for the changes. Looks good to me.

louisfeng committed:
[PyTorch][ET] Collect Process Groups Mapping Info (#104373)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/104373

Add the logics and interface to log ProcessGroup comms configuration (unique ID, type, and ranks info).

We get new information from the ProcessGroup init:
* An unique ID for the PG object
* An unique ID for the actual backend object, this is also exposed in `record_param_comms`, so we can correlate these collectives with the right backend object.
* The config string that shows the backend type (nccl, etc.) and device (cpu, cuda), etc.
* Also provide the global to group ranks mapping.

Test Plan:
Testing in HPC:
```
buck2 build @//mode/opt //hpc/torchrec/models/ads:cmf_10x_launcher --show-output

TORCH_LOGS=all ../buck-out/v2/gen/fbcode/c8344b52091f4f7f/hpc/models/ads/__ads_10x_launcher__/ads_10x_launcher.par  +launcher=local launcher.num_trainers=4 +data_loader=random data_loader.num_batches=2000
```
Example output in ET:
```lang=json
    {
      "name": "## process_group:init ##", "id": 3, "rf_id": 1, "parent": 2, "fw_parent": 0, "seq_id": -1, "scope": 7, "tid": 1, "fw_tid": 0, "op_schema": "",
      "inputs": ["[{\"pg_id\": 139699576920192, \"backend_id\": 139707326969856, \"backend_config\": \"cuda:nccl\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139706539515648, \"backend_id\": 139699576850432, \"backend_config\": \"cuda:nccl\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139706538448512, \"backend_id\": 139706251015872, \"backend_config\": \"cpu:gloo,cuda:gloo\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139706539511808, \"backend_id\": 139689786830656, \"backend_config\": \"cpu:gloo,cuda:gloo\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139700329668864, \"backend_id\": 139689786841856, \"backend_config\": \"cpu:gloo,cuda:gloo\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139700329667328, \"backend_id\": 139689786845440, \"backend_config\": \"cpu:gloo,cuda:gloo\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}]"], "input_shapes": [[]], "input_types": ["String"],
      "outputs": [], "output_shapes": [], "output_types": []
    },
```
Example parser:
```lang=python
import json
json_data = r"""    {
      "name": "## process_group:init ##", "id": 3, "rf_id": 1, "parent": 2, "fw_parent": 0, "seq_id": -1, "scope": 7, "tid": 1, "fw_tid": 0, "op_schema": "",
      "inputs": ["[{\"pg_id\": 139699576920192, \"backend_id\": 139707326969856, \"backend_config\": \"cuda:nccl\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139706539515648, \"backend_id\": 139699576850432, \"backend_config\": \"cuda:nccl\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139706538448512, \"backend_id\": 139706251015872, \"backend_config\": \"cpu:gloo,cuda:gloo\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139706539511808, \"backend_id\": 139689786830656, \"backend_config\": \"cpu:gloo,cuda:gloo\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139700329668864, \"backend_id\": 139689786841856, \"backend_config\": \"cpu:gloo,cuda:gloo\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}, {\"pg_id\": 139700329667328, \"backend_id\": 139689786845440, \"backend_config\": \"cpu:gloo,cuda:gloo\", \"ranks\": {\"0\": 0, \"1\": 1, \"2\": 2, \"3\": 3}}]"], "input_shapes": [[]], "input_types": ["String"],
      "outputs": [], "output_shapes": [], "output_types": []
    }
"""

json_obj = json.loads(json_data)
print(json_obj)
inputs_json = json_obj["inputs"][0]

pg_obj = json.loads(inputs_json)
print(pg_obj)
for pg in pg_obj:
    for key,val in pg.items():
        print(key, val)
```
Full trace on manifold:

```
[trainer2]:I0720 13:01:49.134000 297101 upload_manifold.py:331] PyTorch Execution Trace: manifold hpc_traces/tree/traces/dynocli/devgpu038.ftw6.facebook.com/rank-2/rank-2.Jul_20_13_01_48.297101.et.json.gz
[trainer0]:I0720 13:01:49.133000 307447 upload_manifold.py:331] PyTorch Execution Trace: manifold hpc_traces/tree/traces/dynocli/devgpu038.ftw6.facebook.com/rank-0/rank-0.Jul_20_13_01_48.307447.et.json.gz
[trainer3]:I0720 13:01:49.350000 297112 upload_manifold.py:331] PyTorch Execution Trace: manifold hpc_traces/tree/traces/dynocli/devgpu038.ftw6.facebook.com/rank-3/rank-3.Jul_20_13_01_48.297112.et.json.gz
[trainer1]:I0720 13:01:49.556000 297093 upload_manifold.py:331] PyTorch Execution Trace: manifold hpc_traces/tree/traces/dynocli/devgpu038.ftw6.facebook.com/rank-1/rank-1.Jul_20_13_01_48.297093.et.json.gz
```

Reviewed By: kwen2501, shengbao-zheng

Differential Revision: D46321690

fbshipit-source-id: 500d096e22c1409821fc0eb6660044ed60f5f025

fduwjj commented on torch/distributed/distributed_c10d.py:
```diff
@@ -3895,18 +3923,17 @@ def _new_group_with_tag(
 
     group_name = _process_group_name(ranks, use_hashed_name=use_local_synchronization)
 
-    with record_function(f"## process_group:init with ranks: {ranks}"):
```
Why do we don't need this for profiler?
